File: Java/주민경/[Week 5]완전탐색/[BOJ]블랙잭_2798_브론즈2.java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

class Main {
    public static int max = 0;
    public static void main(String args[]) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String [] input = reader.readLine().split(" ");
        int a = Integer.parseInt(input[0]);
        int b = Integer.parseInt(input[1]);
        int sum = 0;
        /* 두번 째 줄 입력 받아서 리스트로 처리하기 */
        int[] arr = Arrays.stream(reader.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();
        boolean[] visited = new boolean[arr.length];
        combination(arr, visited, 0, a, 3, b);
        System.out.println(max);
    }

    public static void combination(int[] arr, boolean[] visited, int start, int n, int r, int target) {
        if(r == 0) {
            int sum = 0;
            for (int i = 0; i < n; i++) {
                if (visited[i]) {
                    sum += arr[i];
                }
            }
            if ((sum <= target) && (sum > max)) {
                max = sum;
            }
            return;
        } else {
            for(int i = start; i < n; i++) {
                visited[i] = true;
                combination(arr, visited, i + 1, n, r - 1, target);
                visited[i] = false;
            }
        }
    }
}

File: Python/정용준/[Week 5]/[BOJ]블랙잭_2798_브론즈2.py
from itertools  import combinations

# 첫 줄 입력 n, m / n : 카드 개수 m: 목표 숫자
n,m = map(int,input().split())
nums = list(map(int, input().split()))
# 3가지 모든 조합 ([[(5, 6, 7), (5, 6, 8)...])
comb = list(combinations(nums,3))

max_num = 0
#  입력 1에서 모든 조합의 합 (18,19,20,20 ...)
# for j in range(len(comb)):
#     print(sum(comb[j]))


for i in range(len(comb)):
    if(m == sum(comb[i])):
        max_num = sum(comb[i])
    elif (m > sum(comb[i])):
        max_num = max(max_num,sum(comb[i]))
       
print(max_num)